Open Project Dialog

vitvakatu opened the pull request:
### Pull Request Description

Closes #5022 

This is basically a reimplementation of the Open Project Dialog that was present in the IDE a while ago. Now it uses the modern shiny `grid-view` instead of the old rusty `list-view`.

https://user-images.githubusercontent.com/6566674/219052041-ff99aa37-249c-4a63-93a5-5acd6b221dc8.mp4


### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
self-review

farmaazon commented on app/gui/src/presenter/project.rs:
```diff
@@ -336,6 +425,22 @@ impl Project {
         self
     }
 
+    fn attach_frp_to_project_list_notifications(
+        self,
+        notifications: Publisher<ProjectListNotification>,
+        project_list_ready: frp::Source<()>,
+    ) -> Self {
+        let weak = Rc::downgrade(&self.model);
+        let notifications = notifications.subscribe();
+        spawn_stream_handler(weak, notifications, move |notification, _| {
+            match notification {
+                ProjectListNotification::ProjectListReady => project_list_ready.emit(()),
+            }
+            std::future::ready(())
+        });
+        self
+    }
```
I feel it's a bit strange we make a publisher to publish notifications for ourselves. Cannot we just call "reset_entries" on project list in the async block where we wait for new list load?

vitvakatu replied:
Yep, I guess it's a lot easier :) Done.

farmaazon commented on lib/rust/ensogl/core/src/system/js/app.rs:
```diff
@@ -36,6 +36,18 @@ pub mod js_bindings {
         #[wasm_bindgen(method)]
         #[wasm_bindgen(js_name = registerSetShadersRustFn)]
         pub fn register_set_shaders_rust_fn(this: &App, closure: &Closure<dyn FnMut(JsValue)>);
+
+        /// Show a spinner covering the whole viewport.
+        #[allow(unsafe_code)]
+        #[wasm_bindgen(method)]
+        #[wasm_bindgen(js_name = showProgressIndicator)]
+        pub fn show_progress_indicator(this: &App, progress: f32);
+
+        /// Hide a spinner.
+        #[allow(unsafe_code)]
+        #[wasm_bindgen(method)]
+        #[wasm_bindgen(js_name = hideProgressIndicator)]
+        pub fn hide_progress_indicator(this: &App);
```
Somewhere we hide the progress indicator after project load. Can we use this method there? I think I didn't see it in this PR.

vitvakatu replied:
We can't because it is the responsibility of the JS code that loads WASM. Spinner is tracking download progress and is hidden after it is done automatically.

wdanilo replied:
We CAN reuse the spinner if we want. Its API is exposed by JS app that we are connected to in Rust. I can help with it if we want it.

vitvakatu replied:
@wdanilo what do you mean? We *are* using a spinner from JS app, I was just forced to add a bit of code to make it work from Rust (`show_progress_indicator` and `hide_progress_indicator` methods).

vitvakatu committed:
address review

galin-enso commented:
QA passed

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/open-dialog-5022

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/open-dialog-5022

vitvakatu committed:
Close project before opening a new one

vitvakatu commented:
Updated the code to correctly close the project before opening a new one, now, the nodes from the previous project are not visible. The video in the description was updated.

farmaazon commented on app/gui/src/controller/ide.rs:
```diff
@@ -172,6 +174,9 @@ pub trait API: Debug {
     /// Returns `None` if no project is opened at the moment.
     fn current_project(&self) -> Option<model::Project>;
 
+    /// Close the currently opened project. Does nothing if no project is open.
+    fn close_project(&self);
```
This should be part of "ManagingAPI" instead.

farmaazon commented on app/gui/src/presenter.rs:
```diff
@@ -82,6 +82,11 @@ impl Model {
 
+    fn close_project(&self) {
+        *self.current_project.borrow_mut() = None;
+        self.view.project().graph().remove_all_nodes();
```
Add comment why we remove nodes.

MichaelMauderer commented:
**QA**
For me the selection highlight of the project items is not working

https://user-images.githubusercontent.com/1428930/219067687-7f976225-e0db-43c3-be41-7cc07f7f2389.mp4

vitvakatu committed:
Address review comments

vitvakatu commented:
@MichaelMauderer, that's weird, it works for me. 
1. Can you select entries still? 
2. Do you have the grid view demo scene working / other places where grid view is used (CB and its section navigator)?

MichaelMauderer commented:
> Can you select entries still?

Yes, I can select them and opening a new project seems to work.

> Do you have the grid view demo scene working / other places where grid view is used (CB and its section navigator)?

The demo scene and the CB seem to have the correct highlights.

farmaazon commented:
Perhaps the shape ordering is undefined?

vitvakatu committed:
Move grid to different layer

vitvakatu commented:
@MichaelMauderer I moved the grid to the dedicated layer, could you check it?

MichaelMauderer commented:
Still does not appear. There appears to be a shape visible in the debug layers, see the video.


https://user-images.githubusercontent.com/1428930/219346292-700353a0-f5bc-4786-8e1c-7782389cfcd7.mp4


I'm seeing the same behaviour between Chromium and Firefox (both on Linux). I can try a different machine / OS and the electron package if that helps.

wdanilo commented on app/gui/src/presenter/project.rs:
```diff
@@ -243,8 +302,27 @@ impl Project {
+                (rows, cols)
+            }));
+            entry_model <- project_list.grid.model_for_entry_needed.map(f!([model]((row, _)) {
+                model.available_projects.borrow().get(*row).map(|(name, _)| (*row, 0, name.clone_ref()))
```
this is super complex. Also, the `0` is magic

vitvakatu replied:
Fixed

wdanilo reviewed (commented):
Please add spacing in these places: 
<img width="161" alt="image" src="https://user-images.githubusercontent.com/1623053/219904977-c0fabccd-7da8-4cda-8ce1-aeda2726303f.png">

Just like in component browser - the highlights should not touch the borders of the container they are in. Use the same spacing parameters as we use in component browser please.

wdanilo commented:
> Still does not appear. There appears to be a shape visible in the debug layers, see the video.
> 
> 
> I'm seeing the same behaviour between Chromium and Firefox (both on Linux). I can try a different machine / OS and the electron package if that helps.

Regarding this - please remember that if you do not set explicitly the ordering of shapes, it is random and can depend on different factors, liveth order of shapes created, which can be different on different machines. Looks to me like just a lack of shapes ordering specification :)

vitvakatu committed:
Add paddings & fix hover color

vitvakatu commented:
I changed the hover color – the issue was caused by OpenGL bug on my machine that changed the colors for me. Also added paddings as @wdanilo asked, now the window looks like this:

<img width="360" alt="image" src="https://user-images.githubusercontent.com/6566674/220069711-4c459539-6ba0-4110-bb7a-fbe1d34f5c47.png">

(please ignore black hover color, it is an OpenGL bug #4964)

vitvakatu committed:
Fix shapes ordering

MichaelMauderer commented:
QA: Passed.

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/open-dialog-5022

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/open-dialog-5022
